Update stylesheet extraction docs (#1725)
atlassian-labs/compiled: website/packages/docs/src/pages/css-extraction-webpack.mdx

---
section: 50-Guides
name: Stylesheet extraction
---

# Stylesheet Extraction

Stylesheet extraction is **highly recommended** for all component libraries and applications, so that generated class names can be combined and de-duplicated into one external stylesheet. This allows you to enjoy the decreased stylesheet size that comes with atomic styling.

There are two types of stylesheet extraction:

> - Platform stylesheet extraction

Reviewer: Is this relevant for non Atlassians?

Reviewer 2: I would say it is as if say you were a 2nd/3rd-party ecosystem partner using the Atlassian Design System and you are mixing `@atlaskit/button` (post-Compiled-world) with `@your/package` which distributes _runtime styles_ you will experience the same pain we are where now your `border: none` overrides `border-top: 1px solid blue` and your border-top disappears due to runtime injecting a `<style>` tag that results in this being loaded last.

…I think this can go away as a requirement (but technically not a blocker) if we fixed shorthand, @media, and pseudo specificity in other ways, but that's _months off_ even if we knew what that was and started it right away…right?

---
section: 50-Guides
name: Stylesheet extraction
---

# Stylesheet Extraction

Stylesheet extraction is **highly recommended** for all component libraries and applications, so that generated class names can be combined and de-duplicated into one external stylesheet. This allows you to enjoy the decreased stylesheet size that comes with atomic styling.

There are two types of stylesheet extraction:

- Platform stylesheet extraction
- Parcel / Webpack stylesheet extraction

## Which type should I use?

> If your package is a component library, use our **platform stylesheet extraction**. This includes Atlassian Design System packages like `@atlaskit/button`; or any other package that isn't run directly, but is imported and used by _another codebase_ (e.g. Jira, Confluence).

Reviewer 2: Do we want to say "Jira, Confluence" here?  I'm fine with it, but I would probably use more generic wording within the design system:
```suggestion
If your package is a component library, use our **platform stylesheet extraction**. This includes Atlassian Design System packages like `@atlaskit/button` or any other package that is treated as a platform offering or used across products or repositories.
```

Author: ye sure

i went with giving specific atlassian examples because i anticipate that mostly atlassian devs would be reading these docs

but your wording is better

If your package exists in the same codebase as, and is _directly consumed_ by, a Webpack or Parcel app, use **Parcel stylesheet extraction** or **Webpack stylesheet extraction** depending on what bundler your app uses.

Note that Parcel and Webpack are the only bundlers we support.

## Platform stylesheet extraction

Platform stylesheet extraction only works if you use Babel to build your component library. Note that we do not recommend nor support using bundlers (e.g. Webpack or Parcel) to build your component library, because bundlers are appropriate for web apps (e.g. Jira) and not for component libraries.

> **Warning**<br/>
> Ensure that any apps that import your component library also have the Compiled Webpack or Parcel plugins installed (`@compiled/webpack-loader` or `@compiled/parcel-config`), otherwise the order in which the styles are applied will become unpredictable. This can be an issue if you have styles that overlap, e.g. if you mix `padding` and `paddingTop` in the same component, or have `@media` queries that potentially overlap.<br/><br/>

> > If the apps that import your component library use other bundlers like Vite or Next.js, or if your apps cannot install the Compiled Webpack or Parcel plugins, stylesheet extraction will still work! However, you must make sure that within each component, you should _never_ mix any CSS properties or styles that overlap.

Reviewer 2: This will certainly fall on deaf ears, but hey, it's in here!

Author: 🫠

```json
{
  "name": "@example/my-component-library",
  // …
  "dependencies": {
    // Replace "version" with the latest version
    "@compiled/react": "^version"
  },
  // …
  "sideEffects": ["**/*.compiled.css"]
  // …
}
```

To test whether you've set up stylesheet extraction correctly, use your component library's usual build command (e.g. `yarn build`). Inside the `dist` folder, you should see some `.compiled.css` stylesheets generated, one for each `.tsx` file that uses `@compiled/react`.

For example, if you have a file with the path `packages/design-system/button/dist/cjs/button.js` that uses Compiled, you can expect there to be a corresponding `packages/design-system/button/dist/cjs/button.compiled.css`.

## Parcel stylesheet extraction

If you are using a web app that uses Parcel, first install the [Parcel loader](https://compiledcssinjs.com/docs/installation#installation-methods).

Then configure the `extract` plugin to `true` in your Compiled configuration. For example:

```json
// .compiledcssrc
{
  "addComponentName": true,

>   "extract": true, // <-- add this

Reviewer 2: Ah, I should probably add this to `@atlaskit/css`?

Author: hmm i don't know why i kept `inlineCss` in there, i copied this from the jira compiled config 😅 

`addComponentName` is optional, i think it only works with the `styled` API

Reviewer 2: Nevermind, this was already in that documentation, but good catch.

  "inlineCss": true,
  "transformerBabelPlugins": [

>     [
>       "@atlaskit/tokens/babel-plugin",
>       {
>         "shouldUseAutoFallback": true
>       }
>     ]

Reviewer 2: This is default (in the v2 latest version) by the way—we can drop it!
```suggestion
    ["@atlaskit/tokens/babel-plugin"]
```

Author: hmm should it be `"transformerBabelPlugins": [["@atlaskit/tokens/babel-plugin"]]` or `"transformerBabelPlugins": ["@atlaskit/tokens/babel-plugin"]`? something i'll need to double check

Author: never mind, both work https://babeljs.io/docs/plugins/#plugin-options

Reviewer 2: Yeah, _should work_!

+const MiniCssExtractPlugin = require('mini-css-extract-plugin');

module.exports = {
	module: {
		rules: [
			{
				test: /\.(js|ts|tsx)$/,
				exclude: /node_modules/,
				use: [
					{ loader: 'babel-loader' },
					{
						loader: '@compiled/webpack-loader',
						options: {
							extract: true,
						},
					},
				],
			},
+			{

> +				test: /\.compiled-css.css$/i,

Reviewer 2: Not 100% certain on which suggestion it is actually, but isn't it this?
```suggestion
+				test: /^compiled-css\.css$/i,
```
If not, this:
```suggestion
+				test: /\.compiled-css\.css$/i,
```

Author: oops you're right, forgot to update the webpack part

Reviewer 2: Yeah, sounds good.

+				use: [MiniCssExtractPlugin.loader, 'css-loader'],
+			},
			// The following loader will be used for css files that are not from Compiled.
			{
-				test: /\.css$/i,
+				test: /(?<!compiled-css)(?<!\.compiled)\.css$/,
				// Put your existing css loader configuration here
				use: ['style-loader', 'css-loader'],
			},
		],
	},
	plugins: [
+		new MiniCssExtractPlugin(),
		new CompiledExtractPlugin(),
	],
};
```

All extracted styles will be placed in a file called `compiled-css.css`.

> ### Production optimization

Reviewer 2: Do you think I should put this into any of the DEVCOMMs/etc guides?

Author: nah, optional

i don't think anyone has maintained the below Webpack stuff or worked with it for years tbh
